### PR TITLE
SNOW-654808, SNOW-652000: Add field source to basic telemetry data and refactor Telemetry data

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-654808, SNOW-652000

- add source to basic telemetry data
- add a classmethod to `TelemetryData` to construct from a dict

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [x] I am adding a new telemetry message
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [ ] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

**Review by sfc-gh-jdu — APPROVED** [review]
Thanks!

File: src/snowflake/connector/telemetry.py [hunk 1]
```diff
@@ -195,17 +219,25 @@ def buffer_size(self):
         return len(self._log_batch)
 
 
-def generate_telemetry_data(
-    from_dict: dict | None = None, is_oob_telemetry: bool = False
+def generate_telemetry_data_dict(
+    from_dict: dict | None = None,
+    connection: SnowflakeConnection = None,
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/telemetry.py:224]:
```suggestion
    connection: SnowflakeConnection | None = None,
```

File: src/snowflake/connector/telemetry.py [hunk 2]
```diff
@@ -90,6 +91,29 @@ def __init__(self, message, timestamp):
         self.message = message
         self.timestamp = timestamp
 
+    @classmethod
+    def from_telemetry_data_dict(
+        cls,
+        from_dict: dict,
+        timestamp: int,
+        connection: SnowflakeConnection = None,
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/telemetry.py:99]:
```suggestion
        connection: SnowflakeConnection | None = None,
```

**Review by sfc-gh-mkeller — CHANGES_REQUESTED** [review]

**Review by sfc-gh-mkeller — APPROVED** [review]
:ship: :ship:

**Pull request merged**